### PR TITLE
[SPARK-42569][CONNECT] Throw unsupported exceptions for non-supported API

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Match https://github.com/apache/spark/blob/6a2433070e60ad02c69ae45706a49cdd0b88a082/python/pyspark/sql/connect/dataframe.py#L1500 to throw unsupported exceptions in Scala client.

### Why are the changes needed?

Better indicating a API is not supported yet.

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

N/A

### CODE REVIEW

**Comment by amaliujia** [comment]
@hvanhovell

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2461,6 +2461,60 @@ class Dataset[T] private[sql] (val session: SparkSession, private[sql] val plan:
     new DataFrameWriterV2[T](table, this)
   }
 
+  def unpersist(blocking: Boolean): this.type = {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2464]: Did you skip persist for a reason?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2461,6 +2461,60 @@ class Dataset[T] private[sql] (val session: SparkSession, private[sql] val plan:
...
+
+  def localCheckpoint(): Dataset[T] = {
+    throw new UnsupportedOperationException("localCheckpoint is not implemented.")
+  }
+
+  def localCheckpoint(eager: Boolean): Dataset[T] = {
+    throw new UnsupportedOperationException("localCheckpoint is not implemented.")
+  }
+
+  def sameSemantics(other: Dataset[T]): Boolean = {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2506]: We need a story around this one.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2464]: persist is implemented in the Python side I believe

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2506]: Are you saying we should support this?

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2506]: I am saying we need to figure out what to do here.

Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2506]: This seems not to be un-implementable. Maybe deal with it seperately?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2464]: I don't think it is: https://github.com/apache/spark/blob/master/python/pyspark/sql/connect/dataframe.py#L1509-L1510

Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2464]: Oh I now remember:

`def persist(): this.type`
`def persist(newLevel: StorageLevel): this.type`

There is a StorageLevel class in the signature which is from the core module. Does Scala client now can reuse the core 's classes (IIUC scala client is aiming to get rid of core dependency eventually)?

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:2464]: Yeah, we can use the storage enum. We will move the enum to common/util later on.

**Comment by hvanhovell** [comment]
I am merging this one. Can you do persist in a follow-up?

**Pull request closed without merging**